accepted/future-releases/primary-constructors/feature-specification.md

```diff
@@ -575,64 +575,81 @@ constructors as well.
 ```ebnf
 <classDeclaration> ::= // First alternative modified.
      (<classModifiers> | <mixinClassModifiers>)
-     'class' <classNamePart> <superclass>? <interfaces>? <classBody>
+     'class' <classNameMaybePrimary> <superclass>? <interfaces>? <classBody>
    | ...;
 
-<primaryConstructorNoConst> ::= // New rule.
-     <typeIdentifier> <typeParameters>?
-     ('.' <identifierOrNew>)? <declaringParameterList>
+<primaryConstructor> ::= // New rule.
+     'const'? <typeWithParameters> ('.' <identifierOrNew>)?
+     <declaringParameterList>;
 
-<classNamePart> ::= // New rule.
-     'const'? <primaryConstructorNoConst>
+<classNameMaybePrimary> ::= // New rule.
+     <primaryConstructor>
    | <typeWithParameters>;
 
 <typeWithParameters> ::= <typeIdentifier> <typeParameters>?
 
 <classBody> ::= // New rule.
-     '{' (<metadata> <classMemberDeclaration>)* '}'
+     '{' (<metadata> <memberDeclaration>)* '}'
    | ';';
 
 <extensionTypeDeclaration> ::= // Modified rule.
-     'extension' 'type' <classNamePart> <interfaces>?
+     'extension' 'type' <classNameMaybePrimary> <interfaces>?
      <extensionTypeBody>;
 
-<extensionTypeMemberDeclaration> ::= <classMemberDeclaration>;
-
 <extensionTypeBody> ::=
-     '{' (<metadata> <extensionTypeMemberDeclaration>)* '}'
+     '{' (<metadata> <memberDeclaration>)* '}'
    | ';';
 
 <enumType> ::= // Modified rule.
-     'enum' <classNamePart> <mixins>? <interfaces>? '{'
-        <enumEntry> (',' <enumEntry>)* (',')?
-        (';' (<metadata> <classMemberDeclaration>)*)?
+     'enum' <classNameMaybePrimary> <mixins>? <interfaces>? '{'
+        <enumEntry> (',' <enumEntry>)* ','?
+        (';' (<metadata> <memberDeclaration>)*)?
      '}';
 
 <constructorSignature> ::= // Modified rule.
-     <constructorName> <formalParameterList>
+     <constructorName> <formalParameterList> // Old form.
+   | <constructorHead> <formalParameterList> // New form.
    | <declaringConstructorSignature>;
 
 <declaringConstructorSignature> ::= // New rule.
-     'this' ('.' <identifierOrNew>)? <declaringParameterList>?;
+     'this' <identifier>? <declaringParameterList>?;
```

Comment:
I think this should be `'this' (<identifier>? <declaringParameterList>)?`;.

You can't have an identifier without a parameter list.

The `const`, name-identifier and parameter list either all go in the class header or all in an in-body declaration.